tiledb/sm/subarray/range_subset.h

```diff
@@ -0,0 +1,351 @@
+/**
+ * @file   range_subset.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2017-2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the class RangeSetAndSuperset.
+ */
+
+#ifndef TILEDB_RANGE_SUBSET_H
+#define TILEDB_RANGE_SUBSET_H
+
+#include "tiledb/common/heap_memory.h"
+#include "tiledb/sm/enums/datatype.h"
+#include "tiledb/sm/misc/parallel_functions.h"
+#include "tiledb/sm/misc/types.h"
+
+#include <optional>
+#include <type_traits>
+#include <vector>
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
```

Comment:
Don't bother changing it in this PR, but `namespace tiledb::sm { ...` works in C++17. We can update it next time we revisit this code.